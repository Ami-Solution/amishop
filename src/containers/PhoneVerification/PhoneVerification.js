import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import PhoneForm from './PhoneForm';
import ValidateCodeContainer from './ValidateCode';
import CertifyPending from './CertifyPending';
import { LoaderScreen } from '../../components';
import { getErrorMessage } from '../../helpers';
import {
  sendSms as sendSmsAction,
  setPhone as setPhoneAction,
  setPhoneSent as setPhoneSentAction,
  setPhoneCountry as setPhoneCountryAction,
} from '../../actions/kyc';

export class PhoneVerification extends PureComponent {
  static propTypes = {
    isSubmitPhonePending: PropTypes.bool.isRequired,
    sendSms: PropTypes.func.isRequired,
    ethAddress: PropTypes.string.isRequired,
    setPhone: PropTypes.func.isRequired,
    setPhoneCountry: PropTypes.func.isRequired,
    setPhoneSent: PropTypes.func.isRequired,
    phone: PropTypes.string.isRequired,
    phoneCountry: PropTypes.shape({}),
    phoneSent: PropTypes.bool.isRequired,
    phoneVerified: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    phoneCountry: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      phoneNumber: props.phone || '',
      phoneCountry: props.phoneCountry,
      error: '',
      code: null,
    };
  }

  submitPhoneSuccess = ({ data }) => {
    const { phoneNumber, phoneCountry } = this.state;
    const { setPhone, setPhoneSent, setPhoneCountry } = this.props;
    this.setState({ code: data.code });
    this.setState({ lastSend: new Date() });
    setPhone(phoneNumber);
    setPhoneCountry(phoneCountry);
    setPhoneSent(true);
  };

  submitPhoneError = (errors, res) => {
    const message = getErrorMessage(errors, res);
    this.setState({ error: message });
  };

  editPhoneNumber = () => {
    const { setPhoneSent } = this.props;
    setPhoneSent(false);
  };

  submitPhone = (phoneNumber, phoneCountry) => {
    const { sendSms, ethAddress } = this.props;

    const stateToChange = phoneCountry
      ? { phoneNumber, phoneCountry }
      : { phoneNumber };
    this.setState(
      () => stateToChange,
      () => {
        sendSms({
          phoneNumber,
          ethAddress,
          onSuccess: this.submitPhoneSuccess,
          onError: this.submitPhoneError,
        });
      },
    );
  };

  render = () => {
    const { lastSend, phoneNumber, phoneCountry, error, code } = this.state;
    const { phoneSent, phoneVerified, isSubmitPhonePending } = this.props;

    if (phoneVerified) return <CertifyPending />;

    if (!phoneSent && !isSubmitPhonePending) {
      return (
        <PhoneForm
          submitError={error}
          onSubmit={this.submitPhone}
          country={phoneCountry}
        />
      );
    }

    if (isSubmitPhonePending) {
      return <LoaderScreen />;
    }

    return (
      <ValidateCodeContainer
        phoneNumber={phoneNumber}
        editPhoneNumber={this.editPhoneNumber}
        reSendSms={this.submitPhone}
        phoneSent={phoneSent}
        lastSend={lastSend}
        code={code}
      />
    );
  };
}

const mapStateToProps = ({ user, kyc }) => ({
  isSubmitPhonePending: kyc.isSubmitPhonePending,
  ethAddress: user.ethAddress,
  phone: kyc.phone,
  phoneSent: kyc.phoneSent,
  phoneCountry: kyc.phoneCountry,
  phoneVerified: !!kyc.phoneVerified,
});

const mapDispatchToProps = dispatch => ({
  sendSms: params => dispatch(sendSmsAction(params)),
  setPhone: phone => dispatch(setPhoneAction(phone)),
  setPhoneCountry: country => dispatch(setPhoneCountryAction(country)),
  setPhoneSent: bool => dispatch(setPhoneSentAction(bool)),
});

export default connect(mapStateToProps, mapDispatchToProps)(PhoneVerification);
