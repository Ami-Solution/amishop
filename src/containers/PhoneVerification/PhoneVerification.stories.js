/* eslint-disable */
import React from 'react'
import styled, { css } from 'styled-components'
import { storiesOf, addDecorator} from '@storybook/react'
import PhoneVerification from './PhoneVerification'
import { storyWithStore } from '../../helpers'

const initialStore = {
  kyc: {
    isSubmitPhonePending: false,
    isSubmitCodePending: false,
    phone: '',
    phoneSent: false,
    phoneCountry: null,
  },
  user: {
    ethAddress: '0x0000000000000000000000000000000000000000'
  }
}

storiesOf('04 - Screens/OnBoarding', module)
.addDecorator(storyWithStore(initialStore))
.add('Phone - PhoneVerification (all workflow)', () => (
  <PhoneVerification />
))
