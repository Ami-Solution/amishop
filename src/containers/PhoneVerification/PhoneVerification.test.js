import React from 'react';
import { shallow } from 'enzyme';
import { PhoneVerification } from './PhoneVerification';

describe('Containers::PhoneVerification', () => {
  let wrapper;

  beforeEach(() => {
    let requiredProps = {
      isSubmitPhonePending: false,
      sendSms: () => null,
      setPhone: () => null,
      setPhoneCountry: () => null,
      setPhoneSent: () => null,
      ethAddress: '0xawefwefweweq',
      phoneSent: true,
      phoneVerified: false,
      phone: '623423423'
    }
    wrapper = shallow((
      <PhoneVerification
        {...requiredProps}
      />
    ))
  });

  it('should match snapshop', () => {
    expect(wrapper).toMatchSnapshot();
  })
});
