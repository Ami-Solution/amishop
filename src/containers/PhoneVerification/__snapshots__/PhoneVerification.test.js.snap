// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers::PhoneVerification should match snapshop 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhoneVerification
    ethAddress="0xawefwefweweq"
    isSubmitPhonePending={false}
    phone="623423423"
    phoneCountry={null}
    phoneSent={true}
    phoneVerified={false}
    sendSms={[Function]}
    setPhone={[Function]}
    setPhoneCountry={[Function]}
    setPhoneSent={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "code": null,
      "editPhoneNumber": [Function],
      "lastSend": undefined,
      "phoneNumber": "623423423",
      "phoneSent": true,
      "reSendSms": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "code": null,
        "editPhoneNumber": [Function],
        "lastSend": undefined,
        "phoneNumber": "623423423",
        "phoneSent": true,
        "reSendSms": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
